### DISCUSSION TITLE
Is there such util like jotai-reloadable?

### DISCUSSION THREAD

**Discussion by wootra** [pmndrs/jotai#1822] [Ideas] [answered]
Hello, I was trying to make loadable to be re-loaded with simpler way, but couldn't find it. 
I ended up making a new library 'jotai-reloadable'  (https://www.npmjs.com/package/jotai-reloadable). But I think it is better to add this in jotai/util library.
Would you check this and it is worth to add as a util?

**Comment by AjaxSolutions** [comment]
This is where Jotai would nicely integrate with React Query, which has the `refetch` callback.

https://tanstack.com/query/v4/docs/react/reference/useQuery

```
function MyComponent() {
  const [query] = useAtom(queryAtom);
  const { isLoading, error, data, refetch } = useQuery(query.key, () =>
    fetch('/my-endpoint').then(res => res.json()),
    query.options
  );

  return (
    <div>
      {isLoading ? (
        <p>Loading...</p>
      ) : error ? (
        <p>Error: {error.message}</p>
      ) : (
        <div>
          <p>Data: {data}</p>
          <button onClick={refetch}>Refetch</button>
        </div>
      )}
    </div>
  );
}

```

**Reply by wootra** [reply]
I need an atom that holds the result of the service call, and potentially other atoms will derive this atom. with above solution, it will need a component layer to manage an atom and derived atoms are technically depends on the component. What if you forget adding the component and just used the "derived" atoms? 
I don't think above solution is relative to the "loadable" and "reloadable" issue.

**Comment by t6adev** [comment]
If I understand correctly, what you want to do is available through this strategy.
https://jotai.org/docs/recipes/atom-creators#atomwithrefresh

**Reply by wootra** [reply]
> If I understand correctly, what you want to do is available through this strategy. https://jotai.org/docs/recipes/atom-creators#atomwithrefresh

Thanks for the suggestion. But it is a little different issue. I don't want to refresh the service call all the time. I want to reload only when the service call is failed. If I want to force to refetch, I can choose the refresh strategy, but I am pointing out the limitation of loadable utility.
loadable utility gives a good way of handling async processing like react-query as @AjaxSolutions  mentioned. My target is, call the a service for the initial data prefetching, and if some window is popup, refetch only when the pre-fetch is failed. 
if you wrap your atom that holds a Promise with loadable, it will hold the state as { state: 'hasError' }, but it is hard to re-fetch with the structure of `loadable(atom(someAsyncFunc()))`.

**Comment by dai-shi** [comment, marked as answer]
If you want reload in `write` function, it shouldn't be that difficult. If I don't misunderstand:

```js
const fetchFunc = ...
const baseAtom = atom(fetchFunc())
const loadableAtom = loadable(baseAtom)
const finalAtom = atom(
  (get) => get(loadableAtom),
  (get, set) => {
    if (get(loadableAtom).state === 'hasError') {
      set(baseAtom, fetchFunc())
    }
  }
)
```

**Reply by wootra** [reply]
Thanks for answer. The example is what I am currently using. But thought it will be great if the lines can be merged to be 1 line. 
Most of times it should be okay since I don't need to re-load the atom anymore once it succeed, I am a little worried about using loadable util when we want to reload the value with the different arguments (i.e. the service request could have different body). 
when I see loadable code, it is using WeakMap and the key is Promise. Once the new service call is given, the Promise object will be stacked in loadable memo1's weakmap unless the promise-holding atom is changed. 
I think it could go memory leak situation. What do you think?
loadable src : https://github.com/pmndrs/jotai/blob/main/src/vanilla/utils/loadable.ts
where I am concerning:
```
const loadableCache = new WeakMap<Promise<void>, Loadable<Value>>()
const refreshAtom = atom(0)
...
      promise
          .then(
            (data) => {
              loadableCache.set(promise, { state: 'hasData', data })
            },
            (error) => {
              loadableCache.set(promise, { state: 'hasError', error })
            }
```

**Reply by wootra** [reply]
I will take it back. I didn't use weakmap much and was forgetting of the property of the weakmap.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap
By the nature of the weakmap, once the promise is not used anymore, then it will be garbage collected ( If I understand correctly ). I think using loadable is pretty good decision now. 
Thanks for the suggestion and it was a great opportunity for me to learn a good code.
